tests/utils/managed_process.py

```diff
@@ -0,0 +1,239 @@
+        except (OSError, IOError) as e:
+            self._logger.warning(f"Warning: Failed to remove directory {path}: {e}")
+
+    def _check_ports(self, timeout):
+        time_taken = 0
+        for port in self.health_check_ports:
+            time_taken += self._check_port(port, timeout)
+        return time_taken
+
+    def _check_port(self, port, timeout=30, sleep=0.1):
+        """Check if a port is open on localhost."""
+        start_time = time.time()
+        self._logger.info(f"Checking Port: {port}")
+        while time.time() - start_time < timeout:
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                if s.connect_ex(("localhost", port)) == 0:
+                    self._logger.info(f"SUCCESS: Check Port:{port}")
+                    return time.time() - start_time
+            time.sleep(sleep)
+        self._logger.error(f"FAILED: Check Port: {port}")
+        raise RuntimeError(f"FAILED: Check Port: {port}")
+
+    def _check_urls(self, timeout):
+        time_taken = 0
+        for url in self.health_check_urls:
+            time_taken += self._check_url(url, timeout)
+        return time_taken
+
+    def _check_url(self, url, timeout=30, sleep=0.1):
+        if isinstance(url, tuple):
+            response_check = url[1]
+            url = url[0]
+        else:
+            response_check = None
+        start_time = time.time()
+        self._logger.info(f"Checking URL {url}")
+        while time.time() - start_time < timeout:
+            try:
+                response = requests.get(url, timeout=timeout)
+                if response.status_code == 200:
+                    if response_check is None or response_check(response):
+                        self._logger.info(f"SUCCESS: Check URL:{url}")
+                        return time.time() - start_time
+            except requests.RequestException as e:
+                self._logger.warn(f"URL check failed: {e}")
+            time.sleep(sleep)
+
+        self._logger.error(f"FAILED: Check URL: {url}")
+        raise RuntimeError(f"FAILED: Check URL: {url}")
+
+    def _terminate_existing(self):
+        if self.terminate_existing:
+            self._logger.info(f"Terminating Existing {self._command_name}")
+            for proc in psutil.process_iter(["name", "cmdline"]):
+                if proc.name() == self._command_name or proc.name() in self.stragglers:
+                    self._terminate_process_tree(proc.pid)
+
+    def _terminate_process(self, process):
+        try:
+            self._logger.info(f"Terminating {process}")
+            process.terminate()
+        except psutil.AccessDenied:
+            self._logger.warning(f"Access denied for PID {process.pid}")
+        except psutil.NoSuchProcess:
+            self._logger.warning(f"PID {process.pid} no longer exists")
+        except psutil.TimeoutExpired:
+            self._logger.warning(
+                f"PID {process.pid} did not terminate in timeout, killing"
```

Comment:
`"did not terminate in timeout, killing"` => `"did not terminate before timeout, killing"`